lms/static/scripts/frontend_apps/index.tsx

```diff
@@ -39,10 +39,13 @@ export function init() {
   // Construct services used by the file picker app. We need this both for
   // direct launches into this app, and for transitions from viewing to
   // editing assignments.
-  services.set(
-    VitalSourceService,
-    new VitalSourceService({ authToken: config.api.authToken })
-  );
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (config.api?.authToken) {
+    services.set(
+      VitalSourceService,
+      new VitalSourceService({ authToken: config.api.authToken })
+    );
+  }
```

Comment:
I'm not sure what's the right way to handle this.

We have defined the `api` prop to be mandatory in config, but for the email-notifications mode it is not present.

Changing the signature has a lot of implications, so I don't think we want to go that way.

Reply:
The `useConfig` hook has a way to specify that certain optional properties are required, so this means that accurately typing `api` to reflect that it is not present in all modes, is not too difficult.

```diff
diff --git a/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx b/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx
index 4f1cacfb..963ef912 100644
--- a/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx
+++ b/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx
@@ -55,7 +55,7 @@ export default function BasicLTILaunchApp() {
     // Content URL to show in the iframe.
     viaUrl: viaURL,
     canvas,
-  } = useConfig(['hypothesisClient']);
+  } = useConfig(['api', 'hypothesisClient']);
 
   const clientRPC = useService(ClientRPC);
 
diff --git a/lms/static/scripts/frontend_apps/components/ContentSelector.tsx b/lms/static/scripts/frontend_apps/components/ContentSelector.tsx
index 71fab89c..55ca683f 100644
--- a/lms/static/scripts/frontend_apps/components/ContentSelector.tsx
+++ b/lms/static/scripts/frontend_apps/components/ContentSelector.tsx
@@ -91,7 +91,7 @@ export default function ContentSelector({
       },
       youtube: { enabled: youtubeEnabled },
     },
-  } = useConfig(['filePicker']);
+  } = useConfig(['api', 'filePicker']);
 
   // Map the existing content selection to a dialog type and value. We don't
   // open the corresponding dialog immediately, but do pre-fill the dialog
diff --git a/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx b/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx
index 09da619b..3f8b55c1 100644
--- a/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx
+++ b/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx
@@ -105,7 +105,7 @@ export async function loadFilePickerConfig(
     throw new Error('Assignment editing config missing');
   }
 
-  const authToken = config.api.authToken;
+  const authToken = config.api!.authToken;
   const { path, data } = config.editing.getConfig;
   const { assignment, filePicker } = await apiCall<Partial<ConfigObject>>({
     authToken,
@@ -166,7 +166,7 @@ export default function FilePickerApp({ onSubmit }: FilePickerAppProps) {
     },
     assignment,
     filePicker: { deepLinkingAPI, formAction, formFields, promptForTitle },
-  } = useConfig(['filePicker']);
+  } = useConfig(['api', 'filePicker']);
 
   // Currently selected content for assignment.
   const [content, setContent] = useState<Content | null>(
diff --git a/lms/static/scripts/frontend_apps/components/GradingControls.tsx b/lms/static/scripts/frontend_apps/components/GradingControls.tsx
index 3bcf9c9c..7215bf96 100644
--- a/lms/static/scripts/frontend_apps/components/GradingControls.tsx
+++ b/lms/static/scripts/frontend_apps/components/GradingControls.tsx
@@ -39,7 +39,7 @@ export default function GradingControls({
 }: GradingControlsProps) {
   const {
     api: { authToken, sync: syncAPICallInfo },
-  } = useConfig();
+  } = useConfig(['api']);
 
   const clientRPC = useService(ClientRPC);
   const [hasUnsavedChanges, setHasUnsavedChanges] = useState(false);
diff --git a/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx b/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx
index ae48f910..e95d2931 100644
--- a/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx
+++ b/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx
@@ -153,7 +153,7 @@ export default function GroupConfigSelector({
     product: {
       api: { listGroupSets: listGroupSetsAPI },
     },
-  } = useConfig();
+  } = useConfig(['api']);
 
   const useGroupSet = groupConfig.useGroupSet;
   const groupSet = useGroupSet ? groupConfig.groupSet : null;
diff --git a/lms/static/scripts/frontend_apps/config.ts b/lms/static/scripts/frontend_apps/config.ts
index 1452af00..656b45a4 100644
--- a/lms/static/scripts/frontend_apps/config.ts
+++ b/lms/static/scripts/frontend_apps/config.ts
@@ -236,7 +236,9 @@ export type EmailNotificationsPreferences = Record<WeekDay, boolean>;
 export type ConfigObject = {
   // Configuration present in all modes.
   mode: AppMode;
-  api: {
+
+  // Present in most modes.
+  api?: {
     authToken: string;
 
     // Only present in "basic-lti-launch" mode.
diff --git a/lms/static/scripts/frontend_apps/index.tsx b/lms/static/scripts/frontend_apps/index.tsx
index a0853c01..77dffff2 100644
--- a/lms/static/scripts/frontend_apps/index.tsx
+++ b/lms/static/scripts/frontend_apps/index.tsx
@@ -47,7 +47,7 @@ export function init() {
     );
   }
 
-  if (config.rpcServer && config.hypothesisClient) {
+  if (config.api && config.rpcServer && config.hypothesisClient) {
     const { authToken } = config.api;
     const clientRPC = new ClientRPC({
       authToken,
diff --git a/lms/static/scripts/frontend_apps/utils/api.ts b/lms/static/scripts/frontend_apps/utils/api.ts
index abf2d24d..980bc78f 100644
--- a/lms/static/scripts/frontend_apps/utils/api.ts
+++ b/lms/static/scripts/frontend_apps/utils/api.ts
@@ -173,7 +173,7 @@ export function useAPIFetch<T = unknown>(
 ): FetchResult<T> {
   const {
     api: { authToken },
-  } = useConfig();
+  } = useConfig(['api']);
 
   const fetcher: Fetcher<T> | undefined = path
     ? signal =>
```